google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -114,29 +146,56 @@ public boolean next() throws SpannerException {
         @Override
         public void close() {
           super.close();
-          AutoClosingReadContext.this.close();
+          if (isSingleUse) {
+            AutoClosingReadContext.this.close();
+          }
         }
       };
     }
 
+    private void recreateSessionIfPossible(SessionNotFoundException e) {
+      session.maybeRecreateUnderlyingSession(e);
+      session.markUsed();
+      readContextDelegate = readContextDelegateSupplier.get();
```

Comment:
can we just call getReadContextDelegate()?

Reply:
No, it's important that `Supplier.get()` is called here, as the `Supplier` should (have the chance to) re-create the `ReadContext`. If the session was re-created at line 157, the `Supplier` must also re-create the `ReadContext` for it to be associated with the new session.